src/lib/phdi-building-blocks/phdi_building_blocks/schemas.py

```diff
@@ -0,0 +1,251 @@
+import pathlib
+import os
+import yaml
+import json
+import random
+from typing import Literal, List
+import pyarrow as pa
+import pyarrow.parquet as pq
+from fhirpathpy import evaluate
+
+from phdi_building_blocks.fhir import (
+    AzureFhirserverCredentialManager,
+    query_fhir_server,
+    log_fhir_server_error,
+)
+
+
+def load_schema(path: str) -> dict:
+    """
+    Given the path to local YAML files containing a user-defined schema read the file
+    and return the schema as a dictionary.
+
+    :param path: Path specifying the location of a YAML file containing a schema.
+    :return schema: A user-defined schema
+    """
+
+    with open(path, "r") as file:
+        schema = yaml.safe_load(file)
+    return schema
+
+
+def apply_selection_criteria(
+    value: list,
+    selection_criteria: Literal["first", "last", "random", "all"],
+) -> str:
+    """
+    Given a list of value parsed from a FHIR resource and selection criteria return a
+    single value.
+
+    :param value: A list containing the values stored in a top level key in a FHIR
+    resource.
+    :param selection_criteria: A string indicating which element of list to select
+    when one is encountered during parsing.
+    """
+
+    if selection_criteria == "first":
+        value = value[0]
+    elif selection_criteria == "last":
+        value = value[-1]
+    elif selection_criteria == "random":
+        value = random.choice(value)
+    elif selection_criteria == "all":
+        value = value
+
+    # Temporary hack to ensure no structured data is written using pyarrow.
+    # Currently Pyarrow does no support mixing non structure and structured data.
+    # https://github.com/awslabs/aws-data-wrangler/issues/463
+    # Will need to consider other methods of writing to parquet if this is and essential
+    # feature.
+    if type(value) == dict:
+        value = json.dumps(value)
+    elif type(value) == list:
+        value = ",".join(value)
+    return value
+
+
+def apply_schema_to_resource(resource: dict, schema: dict) -> dict:
+    """
+    Given a resource and a schema return a dictionary with values of the data
+    specified by the schema and associated keys defined by the variable name provided
+    by the schema.
+
+    :param resource: A FHIR resource on which to apply a schema.
+    :param schema: A schema specifying the desired values by FHIR resource type.
+    :return data: A dictionary containing the data extracted from the patient a long
+    with specified variable names.
+    """
+
+    data = {}
+    resource_schema = schema.get(resource.get("resourceType", ""))
+    if resource_schema is None:
+        return data
+    for field in resource_schema.keys():
+
+        path = resource_schema[field]["fhir_path"]
+        value = evaluate(resource, path)
+
+        if len(value) == 0:
+            data[resource_schema[field]["new_name"]] = ""
```

Comment:
If `resource_schema[field]["new_name"]` doesn't exist, would it make sense to just default to `field`?  Right now it will throw a NameError I think, which might be OK I guess if we want that to be an error scenario.

Reply:
We know `field` because of how the loop is structured and `"new_name"` should be included in the schema config for every field. I think this is okay and that later we should create a `validate_schema` function that will ensure the user has passed a schema conforming to our required format.